### PR TITLE
Player can move units

### LINKED ISSUE #38: Player can move units
Acceptance criteria:
- [x] select a unit
- [x] mouse cursor changes to 'move here' icon
- [x] when 'move here' icon is shown, and left mouse button is pressed on cell then selected units must move to that cell
- [x] when unit does not face the next 'cell to go to' it should update its facing
- [x] no need for path finding

### PR DESCRIPTION
**Pull request by stefanhendriks**
Fixes #38

- [x] player can select unit
- [x] player can order unit to move
- [x] unit can move pixel by pixel
- [x] unit can have their own move speed
- [x] player can move only its own units
- [x] unit will not stop half-way (pixel based) when ordered to move somewhere else

demo at: https://www.youtube.com/watch?v=lJy1o8QYEhg&feature=youtu.be

After review:
- [x] fix bug when two units move towards each other taking up the same cell (??)
- [x] use polymorphism for `Mouse` class instead of having booleans. (ie, replace procedural polymorphism with inheritance) (demo at: https://www.youtube.com/watch?v=IcIq43m31sE)
- [ ] write/fix tests which are now disabled/removed by refactoring of `Mouse`
- [x] When clicking on another unit - while a unit is already selected - doesn't change the select
- [x] Make `moveSpeed` configurable for `unit` instead of randomized in constructor

### CODE REVIEW

**Comment by stefanhendriks** [comment]
just added - "player can move only its own units"

File: src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java [hunk 1]
```diff
@@ -12,12 +14,18 @@
  */
 public class Mouse {
 
+    private Player playerControlled;
```
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java]: `playerControlled` sounds like a boolean value. I like the name `controllingPlayer` better.

File: src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java [hunk 2]
```diff
@@ -61,13 +72,36 @@ public void deselectEntity() {
...
         }
+        if (selectedEntityBelongsToPlayerControlled()) {
+            movingCursor = false;
+        }
         lastSelectedEntity = null;
     }
 
+    private boolean selectedEntityBelongsToPlayerControlled() {
+        if (lastSelectedEntity == null) return false;
+        if (playerControlled == null) return false;
```
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java]: Maybe I'm a little late to the party, but wouldn't it be better to use polymorphism for `Mouse` to simplify this kind of null-checking logic. I propose to implement 2 concrete mouse classes, e.g. `PlayerControlledMouse` and `HoverCellMouse`. Waddayathink?

**Comment by arjenvanderende** [comment]
When clicking on another unit - while a unit is already selected - doesn't change the select, but moves the selected unit instead. This seems counter-intuitive. Is that supposed to happen?

**Comment by arjenvanderende** [comment]
I'm noticing a z-index issue with the unit selection box: when a unit is selected and it drives diagonally past another unit, then the selection box is draw underneath the other unit. This seems logical: we draw the unit and the selection box in one pass, but they should probably be separate. Either we need to draw selection boxes after all units are rendered, or we need to change the draw order of units so it draws selected units last. I prefer the 1st option.

Should we move this into another issue?

File: src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java [hunk 3]
```diff
@@ -1,41 +1,151 @@
...
+        int drawY = y + offset.getYAsInt();
+        int drawX = x + offset.getXAsInt();
+        graphics.drawImage(sprite, drawX, drawY);
+        this.fadingSelection.render(graphics, drawX, drawY);
+    }
+
+    @Override
+    public void update(float deltaInMs) {
+        this.fadingSelection.update(deltaInMs);
+        if (shouldBeSomewhereElse()) {
```
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java]: Lol, I really like this method name :smile:

File: src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java [hunk 4]
```diff
@@ -1,41 +1,151 @@
...
+            if (isWaitingForNextCellToDetermine()) {
+                decideWhatCellToMoveToNextOrStopMovingWhenNotPossible();
+            } else {
+                // TODO: "make it turn to facing"
```
Comment (arjenvanderende) [inline src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java]: Can we turn this `TODO` into a Github issue/task?

**Comment by arjenvanderende** [comment]
I crashed the game! :smiling_imp:  

To reproduce: place two unit opposite of eachother like this:
screen shot 2015-03-16 at 21 45 23

Then select one unit and move it past the other unit. Quickly deselect the unit and select the other one. Then move that unit in the direction of the other one. Putting the unit further apart makes it easier to do so. If the timings of both units are different (meaning unit 1 is in the center of a cell, while unit 2 is moving between two cells) then they will be allowed to occupy the same cell. After that the game crashes with the following stack trace:
screen shot 2015-03-16 at 21 44 05

File: src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java (hunk 4 above)
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/entities/units/Unit.java]: already done #63

**Comment by stefanhendriks** [comment]
@arjenvanderende created #65

File: src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java (hunk 1 above)
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java]: done

File: src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java (hunk 2 above)
Comment (stefanhendriks) [inline src/main/java/com/fundynamic/d2tm/game/controls/Mouse.java]: Yes mouse states makes sense. Switching between them would be when clicking on a unit or something. Lets do this before merging.

**Comment by stefanhendriks** [comment]
Mouse refactoring is done, still need to add tests, but it feels much better! (it is actually quite easy to add changes to the behaviors now, like for example see what you're placing in debug mode) - 
image

**Comment by stefanhendriks** [comment]
showing off the refactored mouse states at -> https://www.youtube.com/watch?v=IcIq43m31sE 

the only thing you can 'see' is that the mouse no longer snaps to the grid and that you can see what unit/structure is being placed.

**Comment by stefanhendriks** [comment]
Went a bit out of scope, but hey its fun! see: https://www.youtube.com/watch?v=tfRu9bwIoVQ&feature=youtu.be

**Comment by stefanhendriks** [comment]
Bug is fixed!

**Comment by stefanhendriks** [comment]
After watching DHH's talk (https://www.youtube.com/watch?v=9LfmrkyP81M) - I think we should move forward (value) and write tests later. As in, it is a waste to keep this PR open (it is the only active branch, so can be master as well).

**Pull request merged**